sigstore/_internal/sct.py

```diff
@@ -16,75 +16,124 @@
-    cert_der: bytes = cert.public_bytes(encoding=serialization.Encoding.DER)
+    cert_der: bytes = cert.tbs_precertificate_bytes
 
-    # The length should then be split into three bytes.
+    # The length is a u24, which isn't directly supported. So we have to
+    # decompose it into 3 bytes.
     unused, len1, len2, len3 = struct.unpack(
         "!4B",
         struct.pack("!I", len(cert_der)),
     )
     if unused:
         raise InvalidSctError(f"Unexpectedly large certificate length: {len(cert_der)}")
 
+    # No extensions are currently specified.
+    if len(sct.extension_bytes) != 0:
+        raise InvalidSctError("Unexpected trailing extension bytes")
+
+    # NOTE(ww): This is incorrect for non-embedded SCTs, since it unconditionally
+    # embeds the issuer_key_hash.
+
     # Assemble a format string with the certificate length baked in and then pack the digitally
     # signed data
-    pattern = "!BBQhBBB%ssh" % len(cert_der)
+    # fmt: off
+    pattern = "!BBQH32sBBB%ssH" % len(cert_der)
     data = struct.pack(
         pattern,
-        sct.version.value,
-        0,  # Signature Type
-        int(sct.timestamp.timestamp() * 1000),
-        0,  # Entry Type
-        len1,
-        len2,
-        len3,
-        cert_der,
-        len(sct.extensions),
+        sct.version.value,                      # sct_version
+        0,                                      # signature_type (certificate_timestamp(0))
+        int(sct.timestamp.timestamp() * 1000),  # timestamp (milliseconds)
+        sct.entry_type.value,                   # entry_type (x509_entry(0) | precert_entry(1))
+        issuer_key_hash,                        # issuer_key_hash[32]
+        len1,                                   # \
+        len2,                                   # | opaque TBSCertificate<1..2^24-1> OR
+        len3,                                   # | opaque ASN.1Cert<1..2^24-1>
+        cert_der,                               # /
+        len(sct.extension_bytes),               # extensions (opaque CtExtensions<0..2^16-1>)
     )
+    # fmt: on
 
     return data
 
 
+CERTIFICATE_TRANSPARENCY_OID = ObjectIdentifier("1.3.6.1.4.1.11129.2.4.4")
+
+
+def _is_preissuer(issuer: Certificate) -> bool:
+    ext_key_usage = issuer.extensions.get_extension_for_class(ExtendedKeyUsage)
+    return CERTIFICATE_TRANSPARENCY_OID in ext_key_usage.value
+
+
+def _get_issuer_cert(chain: List[Certificate]) -> Certificate:
+    issuer = chain[0]
+    if _is_preissuer(issuer):
+        issuer = chain[1]
+    return issuer
+
+
+def _issuer_key_hash(cert: Certificate) -> bytes:
+    issuer_key: bytes = cert.public_key().public_bytes(
+        encoding=serialization.Encoding.DER,
+        format=serialization.PublicFormat.SubjectPublicKeyInfo,
+    )
+
+    return hashlib.sha256(issuer_key).digest()
+
+
 class InvalidSctError(Exception):
     pass
 
 
 def verify_sct(
-    sct: DetachedFulcioSCT,
+    sct: SignedCertificateTimestamp,
```

Comment:
You've confirmed that the `SignedCertificateTimestamp` structure matches the RFC definition?

Reply:
Yep. `SignedCertificateTimestamp` is `cryptography`'s interface, and they have a large testsuite for conformance.